oak/server/rust/oak_runtime/src/runtime.rs

```diff
@@ -88,6 +98,198 @@ impl Runtime {
+    ///   status for each channel.
+    /// - If all readers are in an erroneous status, e.g. when all `ChannelReaders` are orphaned,
+    ///   this will immediately return the channels statuses.
+    /// - If any of the channels is able to read a message, the corresponding element in the
+    ///   returned vector will be set to `Ok(ReadReady)`, with `Ok(NotReady)` signaling the channel
+    ///   has no message available
+    ///
+    /// In particular, if there is at least one channel in good status and no messages on said
+    /// channel available, `wait_on_channels` will continue to block until a message is
+    /// available.
+    pub fn wait_on_channels(
+        &self,
+        readers: &[Option<&ChannelReader>],
+    ) -> Result<Vec<ChannelReadStatus>, OakStatus> {
+        let thread = oak_platform::current_thread();
+        while !self.is_terminating() {
+            // Create a new Arc each iteration to be dropped after `thread::park` e.g. when the
+            // thread is resumed. When the Arc is deallocated, any remaining `Weak`
+            // references in `Channel`s will be orphaned. This means thread::unpark will
+            // not be called multiple times. Even if thread unpark is called spuriously
+            // and we wake up early, no channel statuses will be ready and so we can
+            // just continue.
+            //
+            // Note we read statuses directly after adding waiters, before blocking to ensure that
+            // there are no messages, after we have been added as a waiter.
+
+            let thread_id = oak_platform::current_thread().id();
+            let thread_ref = Arc::new(thread.clone());
+
+            for reader in readers {
+                if let Some(reader) = reader {
+                    reader.add_waiter(thread_id, &thread_ref);
+                }
+            }
+            let statuses = self.readers_statuses(readers);
+
+            let all_unreadable = statuses.iter().all(|&s| {
+                s == ChannelReadStatus::InvalidChannel || s == ChannelReadStatus::Orphaned
+            });
+            let any_ready = statuses.iter().any(|&s| s == ChannelReadStatus::ReadReady);
+
+            if all_unreadable || any_ready {
+                return Ok(statuses);
+            }
+
+            debug!(
+                "wait_on_channels: channels not ready, parking thread {:?}",
+                oak_platform::current_thread().id()
+            );
+
+            oak_platform::park_thread();
+
+            debug!(
+                "wait_on_channels: thread {:?} re-woken",
+                oak_platform::current_thread().id()
+            );
+        }
+        Err(OakStatus::ErrTerminated)
+    }
+
+    /// Write a message to a channel. Fails with `OakStatus::ErrChannelClosed` if the underlying
+    /// channel has been orphaned.
+    pub fn channel_write(&self, channel: &ChannelWriter, msg: Message) -> Result<(), OakStatus> {
```

Comment:
Being able to access the inner channel is a temporary wart from this incremental approach, although if you are suggesting that a node may try to intentionally bypass checks imposed by the runtime I think this will always be possible. 

Sorry I should have been clearer: this was meant as an incremental approach by first centralising the channel calls and giving the runtime references to all the channels, with follow up PRs addressing tracking of nodes' channels relationships and labels, and for numeric handles. Each of which worth discussing separately, I have an idea for the implementation of the former, but I'm less sure of the need for numeric handles (although this PR does not prevent switching to them). We can discuss more on the numeric handles but my main hesitation is I think it will add indirection without any value- the channel references can be opaque to nodes without being numeric IDs.

Reply:
Got it, thanks for the clarification!

One point:

> We can discuss more on the numeric handles but my main hesitation is I think it will add indirection without any value- the channel references can be opaque to nodes without being numeric IDs.

I would agree, if all nodes would be limited to Rust code, but since at least for wasm they __must__ use numeric IDs, to me it seems fine to focus on that as the only representation. I guess in what you are suggesting, we would use pointers instead of just arbitrary numbers, which I agree seems more idiomatic as long as everything is Rust. Anyways, let's discuss that on a future PR.